### PR TITLE
feat: prioritized nodes

### PR DESCRIPTION
**Pull request by haakonflatval-cognite**
Originally spawned from #1628

### CODE REVIEW

File: viewer/core/src/public/migration/Cognite3DModel.ts [hunk 1]
```diff
@@ -273,7 +273,7 @@ export class Cognite3DModel extends THREE.Object3D implements CogniteModelBase {
    * uses a right-hand Y-up coordinate system. This function also accounts for transformation
    * applied to the model.
    * @param box     The box in ThreeJS/model coordinates.
-   * @param out     Optional preallocated buffer for storing the result. May be `box`.
+   * @param out     Optional preallocated buffer for storing the result. May be same input as  `box`.
```
Comment (larsmoa) [inline viewer/core/src/public/migration/Cognite3DModel.ts:276]:
```suggestion
   * @param out     Optional preallocated buffer for storing the result. May be same input as `box`.
```

File: viewer/index.ts [hunk 2]
```diff
@@ -10,4 +10,5 @@
 
 export * from './core/src/index';
 
-export { NodeAppearance, DefaultNodeAppearance } from './packages/cad-styling';
+export { NodeAppearance, DefaultNodeAppearance, AreaCollection,
+         ClusteredAreaCollection } from './packages/cad-styling';
```
Comment (larsmoa) [inline viewer/index.ts:14]:
```suggestion
export { 
  NodeAppearance, 
  DefaultNodeAppearance, 
  AreaCollection,
  ClusteredAreaCollection
} from './packages/cad-styling';
```

File: viewer/packages/cad-geometry-loaders/src/CadModelUpdateHandler.ts [hunk 3]
```diff
@@ -208,10 +207,12 @@ function createDetermineSectorsInput([settings, camera, clipping, models]: [
   ClippingInput,
   CadNode[]
 ]): DetermineSectorsPayload {
+  const prioritizedAreas = models.map(model => model.prioritizedAreas).flat();
```
Comment (larsmoa) [inline viewer/packages/cad-geometry-loaders/src/CadModelUpdateHandler.ts:210]:
```suggestion
  const prioritizedAreas = models.flatMap(model => model.prioritizedAreas);
```
This is the same, right?

File: viewer/packages/cad-geometry-loaders/src/sector/culling/TakenSectorMap.ts [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+  addSectorCost,
+  SectorLoadingSpent
+} from './types';
+import { CadModelSectorBudget } from '../../CadModelSectorBudget';
+
+import { CadModelMetadata, LevelOfDetail } from '@reveal/cad-parsers';
+
+import assert from 'assert';
+
+export class TakenSectorMap {
```
Comment (larsmoa) [inline viewer/packages/cad-geometry-loaders/src/sector/culling/TakenSectorMap.ts:18]: Pretty sure this file has been moved/renamed and that this class now is unused. Verify and delete this.

File: viewer/packages/cad-geometry-loaders/src/sector/culling/WeightFunctionsHelper.test.ts [hunk 5]
```diff
@@ -5,6 +5,8 @@
 import * as THREE from 'three';
 import { WeightFunctionsHelper } from './WeightFunctionsHelper';
 
+import { PrioritizedArea } from '@reveal/cad-styling';
+
 import { createV8SectorMetadata } from '../../../../../test-utilities';
 
 import { traverseDepthFirst } from '@reveal/utilities';
```
Comment (larsmoa) [inline viewer/packages/cad-geometry-loaders/src/sector/culling/WeightFunctionsHelper.test.ts:12]:
```suggestion
import { PrioritizedArea } from '@reveal/cad-styling';
import { traverseDepthFirst } from '@reveal/utilities';

import { createV8SectorMetadata } from '../../../../../test-utilities';
```

File: viewer/packages/cad-geometry-loaders/src/sector/culling/createDefaultSectorCuller.ts [hunk 6]
```diff
@@ -0,0 +1,14 @@
+/*!
+ * Copyright 2021 Cognite AS
+ */
+import * as THREE from 'three';
+import { SectorCuller } from './SectorCuller';
+import { OccludingGeometryProvider } from './OccludingGeometryProvider';
+import { ByScreenSizeSectorCuller } from './ByScreenSizeSectorCuller';
+
+export function createDefaultSectorCuller(
```
Comment (larsmoa) [inline viewer/packages/cad-geometry-loaders/src/sector/culling/createDefaultSectorCuller.ts:9]: I think this class might have been replaced/removed. Verify and fix.

File: viewer/packages/cad-parsers/src/utilities/types.ts [hunk 7]
```diff
@@ -2,9 +2,10 @@
  * Copyright 2021 Cognite AS
  */
 
-import * as THREE from 'three';
 import { SectorMetadata } from '../metadata/types';
 
+import * as THREE from 'three';
+
```
Comment (larsmoa) [inline viewer/packages/cad-parsers/src/utilities/types.ts:8]:
```suggestion
import * as THREE from 'three';

import { SectorMetadata } from '../metadata/types';


```

File: viewer/packages/cad-styling/src/TreeIndexNodeCollection.ts [hunk 8]
```diff
@@ -2,15 +2,21 @@
  * Copyright 2021 Cognite AS
  */
 import { NumericRange, IndexSet } from '@reveal/utilities';
+import { AreaCollection } from './prioritized/AreaCollection';
+import { ClusteredAreaCollection } from './prioritized/ClusteredAreaCollection';
 import { NodeCollectionBase, SerializedNodeCollection } from './NodeCollectionBase';
 
+import * as THREE from 'three';
+import { EmptyAreaCollection } from './prioritized/EmptyAreaCollection';
```
Comment (larsmoa) [inline viewer/packages/cad-styling/src/TreeIndexNodeCollection.ts:10]:
```suggestion
import * as THREE from 'three';

import { ClusteredAreaCollection } from './prioritized/ClusteredAreaCollection';
import { NodeCollectionBase, SerializedNodeCollection } from './NodeCollectionBase';
import { AreaCollection } from './prioritized/AreaCollection';
import { EmptyAreaCollection } from './prioritized/EmptyAreaCollection';

import { NumericRange, IndexSet } from '@reveal/utilities';
```

File: viewer/packages/cad-styling/src/prioritized/ClusteredAreaCollection.ts [hunk 9]
```diff
@@ -0,0 +1,42 @@
+/*!
+ * Copyright 2021 Cognite AS
+ */
+
+import { AreaCollection } from './AreaCollection';
+import { SmartMergeBoxes } from './mergeboxes/SmartMergeBoxes';
+import { BoxClusterer } from './mergeboxes/BoxClusterer';
```
Comment (larsmoa) [inline viewer/packages/cad-styling/src/prioritized/ClusteredAreaCollection.ts:7]: You don't really need this import - just skip it and omit declaring the type of `_clusterer` below. A bit confusing that there is a interface for this at all - I'd consider just deleting the `BoxClusterer`-interface.

File: viewer/packages/cad-styling/src/prioritized/EmptyAreaCollection.test.ts [hunk 10]
```diff
@@ -0,0 +1,20 @@
...
+import * as THREE from 'three';
+
+describe('EmptyAreaCollection', () => {
+  test('empty area collection does empty area collection stuff', () => {
+    const areaCollection = EmptyAreaCollection.instance();
+
+    expect(areaCollection).toBeTruthy();
+
+    expect(() => areaCollection.addAreas([new THREE.Box3()])).toThrowError();
+    expect(() => areaCollection.intersectWith([new THREE.Box3()])).toThrow();
```
Comment (larsmoa) [inline viewer/packages/cad-styling/src/prioritized/EmptyAreaCollection.test.ts:16]: Should this throw? isn't intersection with an empty area collection just always an empty area collection?

File: viewer/packages/cad-styling/src/prioritized/mergeboxes/SmartMergeBoxes.ts [hunk 11]
```diff
@@ -29,21 +28,36 @@ export class SmartMergeBoxes implements BoxClusterer {
...
+    const size = box.getSize(new Vector3());
+
+    return 2 * (size.x * size.y + size.y * size.z + size.z * size.x);
+  }
+
+  private shouldMergeBoxes(box0: Box3, box1: Box3): boolean {
+    const MAX_SURFACE_INCREASE_RATIO = 1.0;
+    const MAX_SURFACE_INCREASE_ADDITIVE_TERM = 8.0; // Heuristic number of square meters to allow merging of smaller boxes
+
+    const union = box0.clone().union(box1);
```
Comment (larsmoa) [inline viewer/packages/cad-styling/src/prioritized/mergeboxes/SmartMergeBoxes.ts:41]: This causes unnecessary allocations. I suggest you pre-allocate once and reuse for every call. See https://github.com/cognitedata/reveal/blob/9e655cd98775e88e690b517e22f684a62eef91b0/viewer/packages/cad-geometry-loaders/src/sector/culling/OrderSectorsByVisibilityCoverage.ts#L276 for an example.

File: viewer/packages/utilities/src/three/index.ts [hunk 12]
```diff
@@ -1,8 +0,0 @@
-/*!
```
Comment (larsmoa) [inline viewer/packages/utilities/src/three/index.ts:1]: Why was this file removed?

File: viewer/packages/nodes-api/src/NodesCdfClient.ts [hunk 13]
```diff
@@ -76,27 +76,36 @@ export class NodesCdfClient implements NodesApiClient {
...
+    const mappedBoundingBoxes = await Promise.all(mappedBoundingBoxPromises);
+    const resultBoxes = mappedBoundingBoxes
+      .flat()
+      .filter(node => node.boundingBox)
+      .map(node => {
+        const result = new THREE.Box3();
+        const min = node.boundingBox!.min;
+        const max = node.boundingBox!.max;
+        result.min.set(min[0], min[1], min[2]);
+        result.max.set(max[0], max[1], max[2]);
```
Comment (larsmoa) [inline viewer/packages/nodes-api/src/NodesCdfClient.ts:104]: Seems like this could have been replaced with the new `toThreeBox3` function you added in this PR

Comment (larsmoa) [reply viewer/packages/nodes-api/src/NodesCdfClient.ts:104]: viewer/packages/utilities/src/three/toThreeBox3.ts

**Review by larsmoa — CHANGES_REQUESTED** [review]

File: viewer/packages/cad-parsers/src/utilities/types.ts (hunk 7 above)
Comment (haakonflatval-cognite) [reply viewer/packages/cad-parsers/src/utilities/types.ts:8]: Wait, do we have a preferred order for these kinds of imports?? About time I found out, then :sweat_smile:

File: viewer/packages/cad-styling/src/prioritized/ClusteredAreaCollection.ts (hunk 9 above)
Comment (haakonflatval-cognite) [reply viewer/packages/cad-styling/src/prioritized/ClusteredAreaCollection.ts:7]: You're right, BoxClusterer was supposed to be temporary

File: viewer/packages/utilities/src/three/index.ts (hunk 12 above)
Comment (haakonflatval-cognite) [reply viewer/packages/utilities/src/three/index.ts:1]: It's not used anywhere, apparently.. Would there be a point in keeping it?

Comment (larsmoa) [reply viewer/packages/utilities/src/three/index.ts:1]: Ok

File: viewer/packages/cad-styling/src/prioritized/BoxClusterer.ts [hunk 14]
```diff
@@ -29,21 +29,37 @@ export class SmartMergeBoxes implements BoxClusterer {
     this.resultBoxes.pop();
   }
 
-  private shouldMergeBoxesAtIndices(i: number, j: number): boolean {
-    const BOX_MERGE_MIN_IOU = 0.15;
-    const overlap = intersectionOverUnion(this.resultBoxes[i], this.resultBoxes[j]);
-    return overlap >= BOX_MERGE_MIN_IOU;
+  private surfaceArea(box: Box3) {
+    const size = box.getSize(new Vector3());
```
Comment (larsmoa) [inline viewer/packages/cad-styling/src/prioritized/BoxClusterer.ts:33]: Move `new Vector3()` to a pre-allocated var that's reused every call to avoid allocation in this function (like you do with reusableBoundingBox below, but see comments below first).

File: viewer/packages/cad-styling/src/prioritized/BoxClusterer.ts [hunk 15]
```diff
@@ -29,21 +29,37 @@ export class SmartMergeBoxes implements BoxClusterer {
...
+
+    return 2 * (size.x * size.y + size.y * size.z + size.z * size.x);
+  }
+
+  private shouldMergeBoxes(box0: Box3, box1: Box3): boolean {
+    const MAX_SURFACE_INCREASE_RATIO = 1.0;
+    const MAX_SURFACE_INCREASE_ADDITIVE_TERM = 8.0; // Heuristic number of square meters to allow merging of smaller boxes
+
+    const box0Copy = this.reusableBoundingBox.copy(box0);
```
Comment (larsmoa) [inline viewer/packages/cad-styling/src/prioritized/BoxClusterer.ts:42]: See https://github.com/cognitedata/reveal/blob/master/viewer/packages/cad-geometry-loaders/src/sector/culling/OrderSectorsByVisibilityCoverage.ts#L276 for how we've done this before - probably good to keep the pattern

**Review by larsmoa — CHANGES_REQUESTED** [review]

**Review by larsmoa — APPROVED** [review]
